// @nx/react/src/generators/remote/lib/setup-ssr-for-remote.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupSsrForRemote = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nx/devkit");
const versions_1 = require("../../../utils/versions");
function setupSsrForRemote(tree, options, appName) {
    var _a, _b, _c;
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const tasks = [];
        const project = (0, devkit_1.readProjectConfiguration)(tree, appName);
        (0, devkit_1.generateFiles)(tree, (0, devkit_1.joinPathFragments)(__dirname, '../files/module-federation-ssr'), project.root, Object.assign(Object.assign({}, options), { appName, tmpl: '', browserBuildOutputPath: project.targets.build.options.outputPath, serverBuildOutputPath: project.targets.server.options.outputPath }));
        // For hosts to use when running remotes in static mode.
        const originalOutputPath = (_b = (_a = project.targets.build) === null || _a === void 0 ? void 0 : _a.options) === null || _b === void 0 ? void 0 : _b.outputPath;
        project.targets['serve-static'] = {
            dependsOn: ['build', 'server'],
            executor: 'nx:run-commands',
            defaultConfiguration: 'development',
            options: {
                command: `PORT=${(_c = options.devServerPort) !== null && _c !== void 0 ? _c : 4200} node ${(0, devkit_1.joinPathFragments)(originalOutputPath, 'server', 'main.js')}`,
            },
        };
        (0, devkit_1.updateProjectConfiguration)(tree, appName, project);
        const installTask = (0, devkit_1.addDependenciesToPackageJson)(tree, {
            '@module-federation/node': versions_1.moduleFederationNodeVersion,
        }, {});
        tasks.push(installTask);
        return (0, devkit_1.runTasksInSerial)(...tasks);
    });
}
exports.setupSsrForRemote = setupSsrForRemote;
