// @nx/react/src/generators/library/lib/add-linting.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addLinting = void 0;
const tslib_1 = require("tslib");
const linter_1 = require("@nx/linter");
const path_1 = require("nx/src/utils/path");
const json_1 = require("nx/src/generators/utils/json");
const devkit_1 = require("@nx/devkit");
const lint_1 = require("../../../utils/lint");
function addLinting(host, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        if (options.linter === linter_1.Linter.EsLint) {
            const lintTask = yield (0, linter_1.lintProjectGenerator)(host, {
                linter: options.linter,
                project: options.name,
                tsConfigPaths: [
                    (0, path_1.joinPathFragments)(options.projectRoot, 'tsconfig.lib.json'),
                ],
                unitTestRunner: options.unitTestRunner,
                eslintFilePatterns: [`${options.projectRoot}/**/*.{ts,tsx,js,jsx}`],
                skipFormat: true,
                skipPackageJson: options.skipPackageJson,
                setParserOptionsProject: options.setParserOptionsProject,
            });
            (0, json_1.updateJson)(host, (0, path_1.joinPathFragments)(options.projectRoot, '.eslintrc.json'), lint_1.extendReactEslintJson);
            let installTask = () => { };
            if (!options.skipPackageJson) {
                installTask = yield (0, devkit_1.addDependenciesToPackageJson)(host, lint_1.extraEslintDependencies.dependencies, lint_1.extraEslintDependencies.devDependencies);
            }
            return (0, devkit_1.runTasksInSerial)(lintTask, installTask);
        }
        else {
            return () => { };
        }
    });
}
exports.addLinting = addLinting;
