// @nx/react/src/generators/application/lib/add-project.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.maybeJs = exports.addProject = void 0;
const devkit_1 = require("@nx/devkit");
function addProject(host, options) {
    const project = {
        root: options.appProjectRoot,
        sourceRoot: `${options.appProjectRoot}/src`,
        projectType: 'application',
        targets: {},
        tags: options.parsedTags,
    };
    if (options.bundler === 'webpack') {
        project.targets = {
            build: createBuildTarget(options),
            serve: createServeTarget(options),
        };
    }
    (0, devkit_1.addProjectConfiguration)(host, options.projectName, Object.assign({}, project));
}
exports.addProject = addProject;
function maybeJs(options, path) {
    return options.js && (path.endsWith('.ts') || path.endsWith('.tsx'))
        ? path.replace(/\.tsx?$/, '.js')
        : path;
}
exports.maybeJs = maybeJs;
function createBuildTarget(options) {
    var _a;
    return {
        executor: '@nx/webpack:webpack',
        outputs: ['{options.outputPath}'],
        defaultConfiguration: 'production',
        options: {
            compiler: (_a = options.compiler) !== null && _a !== void 0 ? _a : 'babel',
            outputPath: (0, devkit_1.joinPathFragments)('dist', options.appProjectRoot != '.'
                ? options.appProjectRoot
                : options.projectName),
            index: (0, devkit_1.joinPathFragments)(options.appProjectRoot, 'src/index.html'),
            baseHref: '/',
            main: (0, devkit_1.joinPathFragments)(options.appProjectRoot, maybeJs(options, `src/main.tsx`)),
            tsConfig: (0, devkit_1.joinPathFragments)(options.appProjectRoot, 'tsconfig.app.json'),
            assets: [
                (0, devkit_1.joinPathFragments)(options.appProjectRoot, 'src/favicon.ico'),
                (0, devkit_1.joinPathFragments)(options.appProjectRoot, 'src/assets'),
            ],
            styles: options.styledModule || !options.hasStyles
                ? []
                : [
                    (0, devkit_1.joinPathFragments)(options.appProjectRoot, `src/styles.${options.style}`),
                ],
            scripts: [],
            isolatedConfig: true,
            webpackConfig: (0, devkit_1.joinPathFragments)(options.appProjectRoot, 'webpack.config.js'),
        },
        configurations: {
            development: {
                extractLicenses: false,
                optimization: false,
                sourceMap: true,
                vendorChunk: true,
            },
            production: {
                fileReplacements: [
                    {
                        replace: (0, devkit_1.joinPathFragments)(options.appProjectRoot, maybeJs(options, `src/environments/environment.ts`)),
                        with: (0, devkit_1.joinPathFragments)(options.appProjectRoot, maybeJs(options, `src/environments/environment.prod.ts`)),
                    },
                ],
                optimization: true,
                outputHashing: 'all',
                sourceMap: false,
                namedChunks: false,
                extractLicenses: true,
                vendorChunk: false,
            },
        },
    };
}
function createServeTarget(options) {
    return {
        executor: '@nx/webpack:dev-server',
        defaultConfiguration: 'development',
        options: {
            buildTarget: `${options.projectName}:build`,
            hmr: true,
        },
        configurations: {
            development: {
                buildTarget: `${options.projectName}:build:development`,
            },
            production: {
                buildTarget: `${options.projectName}:build:production`,
                hmr: false,
            },
        },
    };
}
