// @nx/react/plugins/with-react.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withReact = void 0;
const processed = new Set();
function addHotReload(config) {
    var _a;
    if (config.mode === 'development' && ((_a = config['devServer']) === null || _a === void 0 ? void 0 : _a.hot)) {
        // add `react-refresh/babel` to babel loader plugin
        const babelLoader = config.module.rules.find((rule) => {
            var _a;
            return rule &&
                typeof rule !== 'string' &&
                ((_a = rule.loader) === null || _a === void 0 ? void 0 : _a.toString().includes('babel-loader'));
        });
        if (babelLoader && typeof babelLoader !== 'string') {
            babelLoader.options['plugins'] = [
                ...(babelLoader.options['plugins'] || []),
                [
                    require.resolve('react-refresh/babel'),
                    {
                        skipEnvCheck: true,
                    },
                ],
            ];
        }
        const ReactRefreshPlugin = require('@pmmmwh/react-refresh-webpack-plugin');
        config.plugins.push(new ReactRefreshPlugin());
    }
}
// We remove potentially conflicting rules that target SVGs because we use @svgr/webpack loader
// See https://github.com/nrwl/nx/issues/14383
function removeSvgLoaderIfPresent(config) {
    const svgLoaderIdx = config.module.rules.findIndex((rule) => typeof rule === 'object' && rule.test.toString().includes('svg'));
    if (svgLoaderIdx === -1)
        return;
    config.module.rules.splice(svgLoaderIdx, 1);
}
/**
 * @param {WithReactOptions} pluginOptions
 * @returns {NxWebpackPlugin}
 */
function withReact(pluginOptions = {}) {
    return function configure(config, context) {
        const { withWeb } = require('@nx/webpack');
        if (processed.has(config))
            return config;
        // Apply web config for CSS, JSX, index.html handling, etc.
        config = withWeb(pluginOptions)(config, context);
        addHotReload(config);
        if ((pluginOptions === null || pluginOptions === void 0 ? void 0 : pluginOptions.svgr) !== false) {
            removeSvgLoaderIfPresent(config);
            config.module.rules.push({
                test: /\.svg$/,
                issuer: /\.(js|ts|md)x?$/,
                use: [
                    {
                        loader: require.resolve('@svgr/webpack'),
                        options: {
                            svgo: false,
                            titleProp: true,
                            ref: true,
                        },
                    },
                    {
                        loader: require.resolve('file-loader'),
                        options: {
                            name: '[name].[hash].[ext]',
                        },
                    },
                ],
            });
        }
        // enable webpack node api
        config.node = {
            __dirname: true,
            __filename: true,
        };
        processed.add(config);
        return config;
    };
}
exports.withReact = withReact;
