// @nx/react/src/rules/update-module-federation-project.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateModuleFederationProject = void 0;
const devkit_1 = require("@nx/devkit");
const versions_1 = require("../utils/versions");
function updateModuleFederationProject(host, options) {
    const projectConfig = (0, devkit_1.readProjectConfiguration)(host, options.projectName);
    projectConfig.targets.build.options = Object.assign(Object.assign({}, projectConfig.targets.build.options), { main: `${options.appProjectRoot}/src/main.ts`, webpackConfig: `${options.appProjectRoot}/webpack.config.js` });
    projectConfig.targets.build.configurations.production = Object.assign(Object.assign({}, projectConfig.targets.build.configurations.production), { webpackConfig: `${options.appProjectRoot}/webpack.config.prod.js` });
    projectConfig.targets.serve.executor =
        '@nx/react:module-federation-dev-server';
    projectConfig.targets.serve.options.port = options.devServerPort;
    // `serve-static` for remotes that don't need to be in development mode
    projectConfig.targets['serve-static'] = {
        executor: '@nx/web:file-server',
        defaultConfiguration: 'development',
        options: {
            buildTarget: `${options.projectName}:build`,
            port: options.devServerPort,
        },
        configurations: {
            development: {
                buildTarget: `${options.projectName}:build:development`,
            },
            production: {
                buildTarget: `${options.projectName}:build:production`,
            },
        },
    };
    (0, devkit_1.updateProjectConfiguration)(host, options.projectName, projectConfig);
    return (0, devkit_1.addDependenciesToPackageJson)(host, {}, { '@nx/web': versions_1.nxVersion });
}
exports.updateModuleFederationProject = updateModuleFederationProject;
